File: server/routes/imports.ts
import { Router, Request, Response, NextFunction } from 'express';
import type { PoolClient } from 'pg';
import { pool } from '../db.js';
import { IMPORT_TYPE_CONFIGS, getImportTypeConfig } from '../importTypes.js';
import { mapAndValidateRows, upsertRows } from '../upsert.js';
import { parseDateOnly } from '../parse.js';
import { isProduction, safeEqual } from '../security.js';

export const importsRouter = Router();

// Autenticação simples por chave compartilhada. Serve tanto para a automação
// Python quanto para a tela de importação do site (mesma chave, injetada no
// frontend via VITE_IMPORT_API_KEY). Se IMPORT_API_KEY não estiver definida,
// a checagem é pulada (conveniente em desenvolvimento local) — nunca deixe
// isso acontecer em produção.
function requireApiKey(req: Request, res: Response, next: NextFunction) {
  const expected = process.env.IMPORT_API_KEY;
  // Sem chave configurada só é tolerado em desenvolvimento local; em produção
  // a rota fica FECHADA (antes ficava aberta a qualquer um).
  if (!expected) {
    if (isProduction) {
      return res.status(503).json({ error: 'IMPORT_API_KEY não configurada no servidor.' });
    }
    return next();
  }
  const provided = req.header('x-api-key') || '';
  if (!safeEqual(provided, expected)) {
    return res.status(401).json({ error: 'Chave de importação ausente ou inválida.' });
  }
  next();
}

interface ImportRequestBody {
  tipo: string;
  dataReferencia: string; // 'YYYY-MM-DD'
  arquivo?: string;
  usuarioNome?: string;
  usuarioEmail?: string;
  mapping: Record<string, string>;
  rows: Record<string, unknown>[];
  /** Índice 0-based do lote fatiado (opcional — default 0). */
  chunkIndex?: number;
  /** Total de lotes desta planilha/aba (opcional — default 1). */
  totalChunks?: number;
  /** id da importação criado no 1º lote — obrigatório a partir do 2º. */
  importId?: number;
  /** Deslocamento 0-based no arquivo original para numerar erros corretamente. */
  rowOffset?: number;
}

/** Limite duro por POST — o frontend fatia planilhas pesadas abaixo deste teto. */
const MAX_ROWS_PER_REQUEST = 2500;

importsRouter.post('/', requireApiKey, async (req: Request, res: Response) => {
  const body = req.body as ImportRequestBody;

  if (!body || typeof body !== 'object') {
    return res.status(400).json({ error: 'Corpo da requisição inválido.' });
  }

  const cfg = getImportTypeConfig(body.tipo);
  if (!cfg) {
    console.error('Unknown import tipo received:', {
      tipo: body.tipo,
      available: Object.keys(IMPORT_TYPE_CONFIGS),
    });
    return res.status(400).json({
      error: `Tipo de importação desconhecido: "${body.tipo}". Tipos válidos: ${Object.keys(IMPORT_TYPE_CONFIGS).join(', ')}.`,
    });
  }

  const dataRefParsed = parseDateOnly(body.dataReferencia);
  if (!dataRefParsed.ok || !dataRefParsed.value) {
    return res.status(400).json({ error: 'data_referencia é obrigatória e deve ser uma data válida (regra 10 do processo).' });
  }
  const dataReferencia = dataRefParsed.value;
  const [anoStr, mesStr] = dataReferencia.split('-');
  const mesReferencia = Number(mesStr);
  const anoReferencia = Number(anoStr);

  if (!Array.isArray(body.rows) || body.rows.length === 0) {
    return res.status(400).json({ error: 'Nenhuma linha para importar.' });
  }
  if (body.rows.length > MAX_ROWS_PER_REQUEST) {
    return res.status(413).json({
      error: `Lote com ${body.rows.length} linhas excede o limite de ${MAX_ROWS_PER_REQUEST} por requisição. Envie a planilha fatiada em partes.`,
    });
  }
  if (!body.mapping || typeof body.mapping !== 'object') {
    return res.status(400).json({ error: 'Mapeamento de colunas ausente.' });
  }

  const chunkIndex = body.chunkIndex ?? 0;
  const totalChunks = body.totalChunks ?? 1;
  if (chunkIndex < 0 || totalChunks < 1 || chunkIndex >= totalChunks) {
    return res.status(400).json({ error: 'Metadados de fatiamento inválidos (chunkIndex/totalChunks).' });
  }
  if (chunkIndex > 0 && (body.importId === undefined || body.importId === null)) {
    return res.status(400).json({ error: 'Lotes seguintes exigem importId retornado no primeiro lote.' });
  }

  const rowOffset = body.rowOffset ?? 0;
  const { valid, errors } = mapAndValidateRows(cfg, body.mapping, body.rows, rowOffset);
  const dataImportacao = new Date();
  const totalLinhas = body.rows.length;
  const clearBefore = chunkIndex === 0;
  const isLastChunk = chunkIndex >= totalChunks - 1;

  // client fora do try (para o catch/finally poderem checar se chegou a
  // existir) — se pool.connect() falhar (ex: banco fora do ar, credencial
  // errada), isso agora vira um 500 tratado em vez de travar a requisição
  // para sempre sem nunca enviar resposta.
  let client: PoolClient | undefined;
  try {
    client = await pool.connect();
    await client.query('BEGIN');

    let importacaoId: number;
    if (chunkIndex === 0) {
      // Reserva o id do log de auditoria antes do upsert, para que cada linha
      // gravada já saia com o importacao_id definitivo (evita um UPDATE extra
      // de correção depois).
      const seqResult = await client.query(
        `SELECT nextval(pg_get_serial_sequence('importacoes', 'id')) AS id`
      );
      importacaoId = Number(seqResult.rows[0].id);
    } else {
      importacaoId = Number(body.importId);
      const existing = await client.query(
        `SELECT id, tipo FROM importacoes WHERE id = $1`,
        [importacaoId]
      );
      if (existing.rowCount === 0) {
        await client.query('ROLLBACK');
        return res.status(400).json({ error: `importId ${importacaoId} não encontrado.` });
      }
      if (String(existing.rows[0].tipo) !== cfg.id) {
        await client.query('ROLLBACK');
        return res.status(400).json({ error: 'importId não corresponde ao tipo desta importação.' });
      }
    }

    const outcome = await upsertRows(
      client,
      cfg,
      valid,
      importacaoId,
      dataImportacao,
      cfg.snapshot ? { dataReferencia, mesReferencia, anoReferencia } : null,
      { clearBefore }
    );

    const status = errors.length === 0 ? 'CONCLUIDO' : valid.length > 0 ? 'CONCLUIDO_COM_AVISOS' : 'FALHA';
    const logStatus =
      !isLastChunk && status !== 'FALHA' ? 'CONCLUIDO_COM_AVISOS' : status;

    if (chunkIndex === 0) {
      await client.query(
        `INSERT INTO importacoes
           (id, tipo, arquivo, usuario_nome, usuario_email, data_referencia, mes_referencia, ano_referencia,
            data_importacao, total_linhas, novos, atualizados, rejeitados, status)
         VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,$10,$11,$12,$13,$14)`,
        [
          importacaoId,
          cfg.id,
          body.arquivo || null,
          body.usuarioNome || null,
          body.usuarioEmail || null,
          dataReferencia,
          mesReferencia,
          anoReferencia,
          dataImportacao,
          totalLinhas,
          outcome.novos,
          outcome.atualizados,
          errors.length,
          logStatus,
        ]
      );
    } else {
      await client.query(
        `UPDATE importacoes
            SET total_linhas = total_linhas + $1,
                novos = novos + $2,
                atualizados = atualizados + $3,
                rejeitados = rejeitados + $4,
                status = $5,
                data_importacao = $6
          WHERE id = $7`,
        [
          totalLinhas,
          outcome.novos,
          outcome.atualizados,
          errors.length,
          isLastChunk ? logStatus : 'CONCLUIDO_COM_AVISOS',
          dataImportacao,
          importacaoId,
        ]
      );
    }

    if (errors.length > 0) {
      const errParams: unknown[] = [];
      const errTuples: string[] = [];
      errors.forEach((e) => {
        const placeholders = [`$${errParams.length + 1}`, `$${errParams.length + 2}`, `$${errParams.length + 3}`];
        errTuples.push(`(${placeholders.join(', ')})`);
        errParams.push(importacaoId, e.linha, e.motivo);
      });
      await client.query(
        `INSERT INTO importacoes_erros (importacao_id, linha, motivo) VALUES ${errTuples.join(', ')}`,
        errParams
      );
    }

    await client.query('COMMIT');

    return res.status(200).json({
      importId: importacaoId,
      tipo: cfg.id,
      tipoLabel: cfg.label,
      arquivo: body.arquivo || null,
      dataReferencia,
      totalAnalisados: totalLinhas,
      novos: outcome.novos,
      atualizados: outcome.atualizados,
      rejeitados: errors.length,
      status,
      erros: errors,
      chunkIndex,
      totalChunks,
      done: isLastChunk,
    });
  } catch (err) {
    if (client) {
      try {
        await client.query('ROLLBACK');
      } catch {
        // se o rollback falhar (ex: conexão já caiu), não há mais nada a fazer aqui
      }
    }
    const message = err instanceof Error ? err.message : 'Erro desconhecido';

    // Mesmo em falha total, registramos a tentativa para auditoria (regra 26),
    // fora da transação que acabou de sofrer rollback.
    try {
      await pool.query(
        `INSERT INTO importacoes
           (tipo, arquivo, usuario_nome, usuario_email, data_referencia, mes_referencia, ano_referencia,
            data_importacao, total_linhas, novos, atualizados, rejeitados, status, mensagem_erro)
         VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9,0,0,$10,'FALHA',$11)`,
        [
          cfg.id,
          body.arquivo || null,
          body.usuarioNome || null,
          body.usuarioEmail || null,
          dataReferencia,
          mesReferencia,
          anoReferencia,
          dataImportacao,
          totalLinhas,
          totalLinhas,
          message,
        ]
      );
    } catch {
      // se nem o log de falha conseguir gravar, não há mais nada a fazer aqui
    }

    // eslint-disable-next-line no-console
    console.error('[imports] Falha na importação:', err);
    return res.status(500).json({ error: 'Falha ao processar a importação. Nenhum dado foi gravado.', detail: message });
  } finally {
    client?.release();
  }
});

importsRouter.get('/', requireApiKey, async (req: Request, res: Response) => {
  try {
    const tipo = typeof req.query.tipo === 'string' ? req.query.tipo : undefined;
    const limit = Math.min(Number(req.query.limit) || 100, 500);

    const params: unknown[] = [];
    let where = '';
    if (tipo) {
      params.push(tipo);
      where = 'WHERE tipo = $1';
    }
    params.push(limit);

    const result = await pool.query(
      `SELECT id, tipo, arquivo, usuario_nome, usuario_email, data_referencia, mes_referencia,
              ano_referencia, data_importacao, total_linhas, novos, atualizados, rejeitados, status, mensagem_erro
       FROM importacoes
       ${where}
       ORDER BY data_importacao DESC
       LIMIT $${params.length}`,
      params
    );

    res.json(result.rows);
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error('[imports] Falha ao consultar histórico:', err);
    res.status(500).json({ error: 'Falha ao consultar o histórico de importações.' });
  }
});

importsRouter.get('/:id/erros', requireApiKey, async (req: Request, res: Response) => {
  try {
    const id = Number(req.params.id);
    if (!Number.isFinite(id)) return res.status(400).json({ error: 'id inválido' });

    const result = await pool.query(
      'SELECT linha, motivo FROM importacoes_erros WHERE importacao_id = $1 ORDER BY linha',
      [id]
    );
    res.json(result.rows);
  } catch (err) {
    // eslint-disable-next-line no-console
    console.error('[imports] Falha ao consultar erros da importação:', err);
    res.status(500).json({ error: 'Falha ao consultar os erros desta importação.' });
  }
});

File: backend/app/routers/imports.py
from __future__ import annotations

from datetime import UTC, datetime
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Path, Query

from app.cache import invalidate_all
from app.db import get_connection
from app.errors import PublicServerError, log_exception, logger
import app.import_types as import_types
from app.parse import parse_date_only
from app.schemas import ImportRequest, ImportResultSummary, ImportRowError
from app.security import require_import_access, require_roles
from app.services import iso
from app.upsert import SnapshotContext, map_and_validate_rows, upsert_rows

router = APIRouter(prefix="/imports", tags=["imports"])


# Limite duro por POST — protege o banco de payloads monstruosos (DoS).
# O frontend fatia planilhas pesadas em lotes menores que este teto.
MAX_ROWS_PER_REQUEST = 2500


@router.post("", response_model=ImportResultSummary)
def create_import(
    body: ImportRequest,
    _access: dict[str, Any] | None = Depends(require_import_access),
) -> ImportResultSummary:
    cfg = import_types.get_import_type_config(body.tipo)
    if not cfg:
        available = ", ".join(sorted(getattr(import_types, "IMPORT_TYPE_CONFIGS", {}).keys()))
        logger.warning("Tipo de importação desconhecido recebido: %r", body.tipo)
        raise HTTPException(
            status_code=400,
            detail=f'Tipo de importação desconhecido: "{body.tipo}". Tipos válidos: {available}.',
        )

    ok, data_referencia = parse_date_only(body.dataReferencia)
    if not ok or not data_referencia:
        raise HTTPException(
            status_code=400,
            detail="data_referencia é obrigatória e deve ser uma data válida (regra 10 do processo).",
        )

    ano_str, mes_str, _ = data_referencia.split("-")
    mes_referencia = int(mes_str)
    ano_referencia = int(ano_str)

    if not body.rows:
        raise HTTPException(status_code=400, detail="Nenhuma linha para importar.")
    if len(body.rows) > MAX_ROWS_PER_REQUEST:
        raise HTTPException(
            status_code=413,
            detail=(
                f"Lote com {len(body.rows)} linhas excede o limite de {MAX_ROWS_PER_REQUEST} "
                "por requisição. Envie a planilha fatiada em partes."
            ),
        )
    if not body.mapping:
        raise HTTPException(status_code=400, detail="Mapeamento de colunas ausente.")

    chunk_index = body.chunkIndex if body.chunkIndex is not None else 0
    total_chunks = body.totalChunks if body.totalChunks is not None else 1
    if chunk_index < 0 or total_chunks < 1 or chunk_index >= total_chunks:
        raise HTTPException(status_code=400, detail="Metadados de fatiamento inválidos (chunkIndex/totalChunks).")
    if chunk_index > 0 and body.importId is None:
        raise HTTPException(
            status_code=400,
            detail="Lotes seguintes exigem importId retornado no primeiro lote.",
        )

    row_offset = body.rowOffset or 0
    valid, errors = map_and_validate_rows(cfg, body.mapping, body.rows, row_offset=row_offset)
    data_importacao = datetime.now(UTC)
    total_linhas = len(body.rows)
    # Só o 1º lote de tipos replace_* pode apagar snapshot/tabela.
    clear_before = chunk_index == 0
    is_last_chunk = chunk_index >= total_chunks - 1

    try:
        with get_connection() as conn:
            if chunk_index == 0:
                seq = conn.execute(
                    "SELECT nextval(pg_get_serial_sequence('importacoes', 'id')) AS id"
                ).fetchone()
                importacao_id = int(seq["id"])
            else:
                importacao_id = int(body.importId)  # type: ignore[arg-type]
                existing = conn.execute(
                    "SELECT id, tipo FROM importacoes WHERE id = %s",
                    (importacao_id,),
                ).fetchone()
                if not existing:
                    raise HTTPException(status_code=400, detail=f'importId {importacao_id} não encontrado.')
                if existing["tipo"] != cfg.id:
                    raise HTTPException(
                        status_code=400,
                        detail="importId não corresponde ao tipo desta importação.",
                    )

            outcome = upsert_rows(
                conn,
                cfg,
                valid,
                importacao_id,
                data_importacao,
                SnapshotContext(data_referencia, mes_referencia, ano_referencia) if cfg.snapshot else None,
                clear_before=clear_before,
            )

            status_label = (
                "CONCLUIDO"
                if not errors
                else "CONCLUIDO_COM_AVISOS"
                if valid
                else "FALHA"
            )
            # Enquanto houver lotes pendentes, marca como parcial no log.
            if not is_last_chunk and status_label != "FALHA":
                log_status = "CONCLUIDO_COM_AVISOS"
            else:
                log_status = status_label

            if chunk_index == 0:
                conn.execute(
                    """
                    INSERT INTO importacoes
                      (id, tipo, arquivo, usuario_nome, usuario_email, data_referencia, mes_referencia, ano_referencia,
                       data_importacao, total_linhas, novos, atualizados, rejeitados, status)
                    VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s,%s)
                    """,
                    (
                        importacao_id,
                        cfg.id,
                        body.arquivo,
                        body.usuarioNome,
                        body.usuarioEmail,
                        data_referencia,
                        mes_referencia,
                        ano_referencia,
                        data_importacao,
                        total_linhas,
                        outcome.novos,
                        outcome.atualizados,
                        len(errors),
                        log_status,
                    ),
                )
            else:
                conn.execute(
                    """
                    UPDATE importacoes
                       SET total_linhas = total_linhas + %s,
                           novos = novos + %s,
                           atualizados = atualizados + %s,
                           rejeitados = rejeitados + %s,
                           status = %s,
                           data_importacao = %s
                     WHERE id = %s
                    """,
                    (
                        total_linhas,
                        outcome.novos,
                        outcome.atualizados,
                        len(errors),
                        log_status if is_last_chunk else "CONCLUIDO_COM_AVISOS",
                        data_importacao,
                        importacao_id,
                    ),
                )

            if errors:
                with conn.cursor() as cur:
                    cur.executemany(
                        "INSERT INTO importacoes_erros (importacao_id, linha, motivo) VALUES (%s, %s, %s)",
                        [(importacao_id, e.linha, e.motivo) for e in errors],
                    )

            # commit final import metadata (upsert_rows commits per chunk)
            conn.commit()
            invalidate_all()

        return ImportResultSummary(
            importId=importacao_id,
            tipo=cfg.id,
            tipoLabel=cfg.label,
            arquivo=body.arquivo,
            dataReferencia=data_referencia,
            totalAnalisados=total_linhas,
            novos=outcome.novos,
            atualizados=outcome.atualizados,
            rejeitados=len(errors),
            status=status_label,  # type: ignore[arg-type]
            erros=[ImportRowError(linha=e.linha, motivo=e.motivo) for e in errors],
            chunkIndex=chunk_index,
            totalChunks=total_chunks,
            done=is_last_chunk,
        )
    except HTTPException:
        raise
    except Exception as err:
        # Detalhe técnico só no log do servidor; na tela/histórico fica a mensagem
        # genérica com o código de referência para o suporte localizar o registro.
        ref = log_exception(f"Falha na importação tipo={cfg.id} arquivo={body.arquivo!r}", err)
        message = f"Falha interna ao processar a importação (ref. {ref})."
        try:
            with get_connection() as conn:
                conn.execute(
                    """
                    INSERT INTO importacoes
                      (tipo, arquivo, usuario_nome, usuario_email, data_referencia, mes_referencia, ano_referencia,
                       data_importacao, total_linhas, novos, atualizados, rejeitados, status, mensagem_erro)
                    VALUES (%s,%s,%s,%s,%s,%s,%s,%s,%s,0,0,%s,'FALHA',%s)
                    """,
                    (
                        cfg.id,
                        body.arquivo,
                        body.usuarioNome,
                        body.usuarioEmail,
                        data_referencia,
                        mes_referencia,
                        ano_referencia,
                        data_importacao,
                        total_linhas,
                        total_linhas,
                        message,
                    ),
                )
                conn.commit()
        except Exception:
            pass
        raise PublicServerError("Falha ao processar a importação. Nenhum dado foi gravado.", ref=ref) from err


@router.get("")
def list_imports(
    tipo: str | None = Query(default=None, max_length=64, pattern=r"^[A-Za-z0-9_\-]+$"),
    limit: int = Query(default=100, ge=1, le=500),
    _admin: dict[str, Any] = Depends(require_roles("ADMIN")),
) -> list[dict[str, Any]]:
    params: list[Any] = []
    where = ""
    if tipo:
        where = "WHERE tipo = %s"
        params.append(tipo)
    params.append(limit)
    with get_connection() as conn:
        rows = conn.execute(
            f"""
            SELECT id, tipo, arquivo, usuario_nome, usuario_email, data_referencia, mes_referencia,
                   ano_referencia, data_importacao, total_linhas, novos, atualizados, rejeitados, status, mensagem_erro
            FROM importacoes
            {where}
            ORDER BY data_importacao DESC
            LIMIT %s
            """,
            params,
        ).fetchall()
    out: list[dict[str, Any]] = []
    for r in rows:
        item = dict(r)
        item["data_referencia"] = iso(item.get("data_referencia"))
        item["data_importacao"] = iso(item.get("data_importacao"))
        out.append(item)
    return out


@router.delete("/history")
def clear_import_history(
    _admin: dict[str, Any] = Depends(require_roles("ADMIN")),
) -> dict[str, Any]:
    """Remove todo o histórico de importações (e erros associados via CASCADE)."""
    with get_connection() as conn:
        # Apaga erros primeiro para cobrir FKs sem ON DELETE CASCADE em bases legadas.
        err = conn.execute("DELETE FROM importacoes_erros")
        result = conn.execute("DELETE FROM importacoes")
        conn.commit()
        deleted = int(result.rowcount or 0)
        deleted_errors = int(err.rowcount or 0)
    return {"ok": True, "deleted": deleted, "deletedErrors": deleted_errors}


@router.get("/{import_id}/erros")
def list_import_errors(
    import_id: int = Path(ge=1, le=2**31 - 1),
    _admin: dict[str, Any] = Depends(require_roles("ADMIN")),
) -> list[dict[str, Any]]:
    with get_connection() as conn:
        rows = conn.execute(
            "SELECT linha, motivo FROM importacoes_erros WHERE importacao_id = %s ORDER BY linha",
            (import_id,),
        ).fetchall()
    return [dict(r) for r in rows]
